[docs] Add docs for react navigation integration (#46319)
expo/expo: docs/pages/eas/observe/integrations/react-navigation.mdx

---
title: React Navigation integration
sidebar_title: React Navigation
description: Track per-screen render and interactive timings by enabling the React Navigation integration for EAS Observe.
---

import { Prerequisites, Requirement } from '~/ui/components/Prerequisites';
import { Step } from '~/ui/components/Step';

EAS Observe ships an opt-in integration for [React Navigation](https://reactnavigation.org/) that collects per-screen metrics tagged with the screen's path (for example, `/sessions/abc`). This lets you compare navigation performance by screen in the dashboard instead of looking only at app-wide aggregates.

If your app uses [Expo Router](/router/introduction/), use the [Expo Router integration](/eas/observe/integrations/expo-router/) instead. This page is for apps that use React Navigation directly.

## Prerequisites

<Prerequisites>

>   <Requirement title="Expo SDK 56 or later">

Reviewer: **[nit]** The glossary prefers `SDK 56 and later` for version ranges, and the body of this same `Requirement` (line 18) already uses `SDK 56 and later`. Align the title so the wording matches (`expo-docs-style-guide.md#glossary-and-product-names`).

```suggestion
  <Requirement title="Expo SDK 56 and later">
```

    The React Navigation integration is available on SDK 56 and later. On earlier SDKs,
    `expo-observe` still tracks app-wide metrics, but per-screen navigation events are not emitted.
  </Requirement>
  <Requirement title="An app already using EAS Observe">
    Follow [Get started](/eas/observe/get-started/) to install `expo-observe` and create your first
    build.
  </Requirement>

>   <Requirement title="@react-navigation/native 7 installed in the app">

Reviewer: **[nit]** The title says `@react-navigation/native 7`, but the body (line 26) clarifies `v7.0.0 or later`. The title reads as if v7 exactly is required. Match the body's intent so the requirement is unambiguous in the heading map (`expo-docs-style-guide.md#glossary-and-product-names`).

```suggestion
  <Requirement title="@react-navigation/native 7 or later installed in the app">
```

    The integration depends on `@react-navigation/native` (v7.0.0 or later) at runtime. If the
    package is not installed, the integration becomes a silent no-op.
  </Requirement>
</Prerequisites>

<Step label="1">

## Enable the integration

> **warning** The integration must be enabled before mount and cannot be toggled at runtime. Calling `configure()` after the app has mounted, or toggling the flag mid-session, throws an error.

Call `Observe.configure()` with the `react-navigation` integration flag at module scope, before any screen mounts:

```tsx App.tsx
import { Observe } from 'expo-observe';

Observe.configure({
  integrations: { 'react-navigation': true },
});
```

</Step>

<Step label="2">

## Use `ObserveNavigationContainer`

> Replace your top-level `<NavigationContainer>` with `<ObserveNavigationContainer>`. It wraps the stock container, accepts the same props, and forwards the same ref, while subscribing to navigation state changes so it can record per-screen render timings.

Reviewer: **[suggestion]** The second sentence is 32 words and stacks three coordinated verb phrases (`wraps`, `accepts`, `forwards`) plus a `while` clause. Splitting it puts the subscription behavior on its own beat (`expo-docs-style-guide.md#voice-and-tone`).

```suggestion
Replace your top-level `<NavigationContainer>` with `<ObserveNavigationContainer>`. It wraps the stock container, accepts the same props, and forwards the same ref. It also subscribes to navigation state changes so it can record per-screen render timings.
```

```tsx App.tsx
import { Observe } from 'expo-observe';
import { ObserveNavigationContainer } from 'expo-observe/integrations/react-navigation';

Observe.configure({
  integrations: { 'react-navigation': true },
});

export default function App() {

>   return <ObserveNavigationContainer>{/* your navigators */}</ObserveNavigationContainer>;

Reviewer 2: nit: Should we format this with line breaks?

```suggestion
  return (
    <ObserveNavigationContainer>
      {/* your navigators */}
    </ObserveNavigationContainer>
  );
```

</Step>

## Metrics

### Per-screen first render (`cold_ttr`)

**What it measures:** Time from when a navigation action is dispatched (for example, `navigation.navigate()`) to when the destination screen first becomes focused. For the very first focus after app launch, the measurement is taken from when the JS bundle is loaded, and the event includes `isAppLaunch: true`.

Emitted at most once per screen instance within a session.

**Event params:**

| Param         | Type      | Description                                                                                             |
| ------------- | --------- | ------------------------------------------------------------------------------------------------------- |
| `routeName`   | `string`  | Resolved screen path (for example `/sessions/abc`), or `route.name` if no `linking` config resolves it. |
| `routeParams` | `object`  | Focused route params (for example, `{ sessionId: 'abc' }`).                                             |
| `isAppLaunch` | `boolean` | `true` when measured against process start, `false` for subsequent navigation.                          |

### Per-screen warm render (`warm_ttr`)

> **What it measures:** Same as `cold_ttr`, but for screens that were already rendered before focus, typically because they were preloaded or the user navigated back to them. Tab-navigator siblings count as warm only once they have been mounted; with React Navigation v7's default `lazy: true`, unfocused tabs stay unmounted and their first focus is recorded as `cold_ttr`.

Reviewer: **[suggestion]** Three ideas joined with a semicolon and a comma: tab-navigator warm semantics, React Navigation's `lazy: true` default, and the resulting `cold_ttr` recording. The semicolon is doing the work of a sentence break (`expo-docs-style-guide.md#punctuation`).

```suggestion
**What it measures:** Same as `cold_ttr`, but for screens that were already rendered before focus, typically because they were preloaded or the user navigated back to them. Tab-navigator siblings count as warm only once they have been mounted. With React Navigation v7's default `lazy: true`, unfocused tabs stay unmounted and their first focus is recorded as `cold_ttr`.
```

**Event params:**

| Param         | Type     | Description                                                 |
| ------------- | -------- | ----------------------------------------------------------- |
| `routeName`   | `string` | Resolved screen path, or `route.name`.                      |
| `routeParams` | `object` | Focused route params (for example, `{ sessionId: 'abc' }`). |

### Per-screen time to interactive (`tti`)

**What it measures:** Time from when a navigation action is dispatched to when `markInteractive()` is called on the destination screen. Only the first call per navigation is recorded, so it is safe to call `markInteractive()` multiple times.

**Event params:**

| Param         | Type     | Description                                              |
| ------------- | -------- | -------------------------------------------------------- |
| `routeName`   | `string` | Resolved screen path, or `route.name`.                   |
| `routeParams` | `object` | Focused route params.                                    |
| `...`         | `any`    | Any custom params passed via `markInteractive({ ... })`. |

## Notes and troubleshooting

> - The integration only activates if `@react-navigation/native` is installed at runtime. If it is not installed, `useObserve()` continues to work but no per-screen navigation metrics are emitted, and rendering `<ObserveNavigationContainer>` throws.

Reviewer: **[suggestion]** Three clauses chained with `but` and `and` make the failure case (rendering `<ObserveNavigationContainer>` throws) easy to miss on a skim (`expo-docs-style-guide.md#voice-and-tone`).

```suggestion
- The integration only activates if `@react-navigation/native` is installed at runtime. If it is not installed, `useObserve()` continues to work, but no per-screen navigation metrics are emitted. Rendering `<ObserveNavigationContainer>` without `@react-navigation/native` throws.
```
